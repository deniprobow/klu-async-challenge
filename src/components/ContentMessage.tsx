import React, { ReactElement, useCallback, useState } from 'react'
import styles from '../styles/ContentMessage.module.css'
import {TimeHourMinuteFormatter} from '../utils/DateTimeFormatter'
import { FaTrash } from "react-icons/fa";
import { useMessage } from '@/context/MessageContext';
import LinkPreview from './LinkPreview';
import Modal from './Modal';
import { Anchorme } from 'react-anchorme'


function ContentMessage(props:any) {
    const { message, deleteMessage } = useMessage();
    const [showingModal, setShowingModal] = useState(false);
    const [idMessage, setIdMessage] = useState('');
    
    const handleDelete = useCallback((id:string) =>{ 
        setShowingModal(true)
        setIdMessage(id);
    },[message, showingModal, idMessage])

    const closeModal = ()=> setShowingModal(false)
    
    return (
        <>
        <div className={styles.contentMessage} data-testid={'list-content-message'}>
            {
                props.datas &&
                props.datas.sort() &&
                props.datas.map((el:any)=>{
                    return(
                        <div key={el.id} >
                        <div className={styles.detailMessage} >
                            <Anchorme data-id={el.id}>{el.message}</Anchorme><br/>
                            <small className={styles.detailTime}>{TimeHourMinuteFormatter(el.datetime)}</small> <button onClick={()=>handleDelete(el.id)} data-testid={el.message} className={styles.deleteMessageIconButton} > <span className={styles.deleteMessageIcon}><FaTrash /></span> </button>
                        </div>
                        {
                            el.message.match(/\bhttps?:\/\/\S+/gi) &&
                            (
                                <LinkPreview
                                    url={el.message.match(/\bhttps?:\/\/\S+/gi)} // link you want to preview
                                />
                            )
                        } 
                        </div>
                        )
                    
                })
            }
        </div>
        {
            showingModal &&
            <Modal idMessage={idMessage} title='Confirm' description='Delete the message?' deleteMessage={deleteMessage} closeModal={closeModal}  />
        }
        </>
    )
}
export default ContentMessage
