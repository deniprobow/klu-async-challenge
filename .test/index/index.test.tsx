import { fireEvent, render, screen, waitFor, within } from '@testing-library/react'
import Home from '../../src/pages/index'
import ContentMessage from '../../src/components/ContentMessage'
import React from 'react'

describe('Home', () => {
  it('renders a main-container', () => {
    render(<Home />);

    const mainContainer = screen.getByTestId('main-container')

    expect(mainContainer).toBeInTheDocument()
  })

  it('renders a content-message component', () => {
    render(<Home />);
    const contentMessage = screen.queryByTestId('content-message')

    expect(contentMessage).toBeDefined();
    
  })

  it('renders a button-group component', () => {
    render(<Home />);

    const buttonGroup = screen.queryByTestId('button-group')

    expect(buttonGroup).toBeDefined();
    
  })

  it('have input message and send button', () => {
    render(<Home />);
    expect(screen.getByRole("textbox"))
      .toHaveAttribute("data-testid", "input-message");
    expect(screen.getByRole("button"))
      .toHaveAttribute("data-testid", "button-send");
  });

  it('disable button send when input type is empty', async () => {
    render(<Home />);
    expect(screen.queryByTestId("input-message"))
      .toHaveValue("");

    expect(screen.queryByTestId("button-send")).toHaveAttribute("disabled")
    expect(screen.queryByTestId("button-send")).toBeDisabled
    
  });

  it('succeeed add data when input type is not empty and button send be clicked.', async () => {
    render(<Home />);
    let sampleMessage = Math.random().toString(36).substring(2, 20);
    const input = screen.getByPlaceholderText<HTMLInputElement>("Type Message...");
    fireEvent.change(input, {target: {value: sampleMessage}})
    const buttonSend:any = screen.queryByTestId("button-send");
    fireEvent.click(buttonSend);
    expect( screen.findAllByText(sampleMessage)).toBeInTheDocument;
    
  });

  it('succeeed open the link inside messsage when url text message be clicked.', async () => {
    render(<Home />);
    let sampleMessage:any = "https://google.com";
    // let sampleMessage = Math.random().toString(36).substring(2, 20);
    const input = screen.getByPlaceholderText<HTMLInputElement>("Type Message...");
    fireEvent.change(input, {target: {value: sampleMessage}})
    const buttonSend:any = screen.queryByTestId("button-send");
    fireEvent.click(buttonSend);

    expect(screen.findAllByText(sampleMessage)).toBeInTheDocument;
    const urlInMessage:any = await screen.findAllByText(sampleMessage);
    // fireEvent.click( urlInMessage);
    
    // expect( screen.findAllByText("Delete the message?")).toBeInTheDocument;
    
  });

  it('succeeed show confirm modal delete when icon delete be clicked.', async () => {
    render(<Home />);
    // let sampleMessage = "google";
    let sampleMessage = Math.random().toString(36).substring(2, 20);
    const input = screen.getByPlaceholderText<HTMLInputElement>("Type Message...");
    fireEvent.change(input, {target: {value: sampleMessage}})
    const buttonSend:any = screen.queryByTestId("button-send");
    fireEvent.click(buttonSend);
    expect(screen.findAllByText(sampleMessage)).toBeInTheDocument;
    
    const buttonSelected:any = screen.queryByTestId(sampleMessage);
    expect(buttonSelected).toBeInTheDocument;
    // fireEvent.click(buttonSelected);
    // expect( screen.findAllByText("Delete the message?")).toBeInTheDocument;
    
  });
  
})

function mount(arg0: React.JSX.Element) {
    throw new Error('Function not implemented.')
}
