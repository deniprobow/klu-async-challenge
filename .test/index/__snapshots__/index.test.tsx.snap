// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders a button-group component 1`] = `null`;

exports[`Home renders a content-message 1`] = `null`;

exports[`Home renders a content-message 2`] = `null`;

exports[`Home renders a content-message component 1`] = `null`;

exports[`Home renders a input-message component 1`] = `null`;

exports[`Home renders a main-container 1`] = `
<main
  class="wrapper"
  data-testid="main-container"
>
  <div
    class="contentMessage"
  />
  <div
    class="inputGroup"
  >
    <input
      class="inputMessage"
      placeholder="Type Message..."
      value=""
    />
    <button
      class="buttonSend"
      type="button"
    >
      Send
      <svg
        class="buttonSendIcon"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 19l9 2-9-18-9 18 9-2zm0 0v-8"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
  </div>
</main>
`;
